src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -413,6 +421,10 @@ MulticopterPositionControl::poll_subscriptions()
 		if (_wv_controller == nullptr && _vehicle_status.is_vtol) {
 			_wv_controller = new WeatherVane();
 		}
+		// if the vehicle is a rotary wing, enable collision avoidance capabilities
+		if (_ca_controller == nullptr && _vehicle_status.is_rotary_wing) {
+			_ca_controller = new CollisionAvoidance();
```

Comment:
@MaEtUgR The weathervane controller is only allocated is the vehicle is a vtol. A vehicle will not all of a sudden become a vtol when it's in air so there should be no reason for concern.
The reason why myself and @dagar decided to do it this way is firstly because it avoids adding weathervane parameters for non-vtol vehicles. And secondly there is currently no other interface to add features that are not limited to just the flighttask interface.

Reply:
Thanks for your answer. Please let me clarify my thoughts:

> adding weathervane parameters for non-vtol vehicles
I clearly see this point but there is hopefully a better solution.

> A vehicle will not all of a sudden become a vtol when it's in air so there should be no reason for concern.
- Can you guarantee that for any future change or bug?
- You are checking if it suddenly became a vtol on every loop iteration. So if there was any bug you would certainly catch it also in air.

> The weathervane controller is only allocated if the vehicle is a vtol.
The ideology behind PX4 is that a board supports all the compiled features with the available memory, right?. So I don't see a point in dynamic RAM saving dependent on configuration, it might lead to unforseen memory overflows on certain boards with certain configurations. I would support a fine grained choice of features on compile time because one board doing everything without reflash is not the only PX4 application. But that would require a change in ideology.

If you still want to dynamically allocate weathervane then please do that in the most safe way you can think of and it should not be the standard to introduce new libraries. If there's flaws we need to improve the software architecture.